Add (un)subscription to PluginContext (#1792)
kuzzleio/kuzzle: doc/2/plugins/plugin-context/accessors/realtime/index.md

---
code: true
type: page

> title: registerSubscription

Reviewer: ```suggestion
title: realtime
```

---

# `realtime`

> Adds or removes [real-time subscriptions](/core/2/guides/essentials/real-time) from the backend.

Reviewer 2: ```suggestion
Manage [real-time subscriptions](/core/2/guides/essentials/real-time) from the backend.
```
For now it's only add/remove but it may evolve

Reviewer: (Manages)

## `subscribe`

> Registers a new realtime subscription on behalf of a client. The subscription works exactly like the one created by the [`realtime:subscribe` API endpoint](/api/controllers/realtime/subscribe/index.md). The notifications will be sent to the connection identified by the `Request` object passed to the method.

Reviewer: This method documentation is obsolete: it doesn't accept a `Request` object anymore.

> Registers a new realtime subscription on behalf of a client. The subscription works exactly like the one created by the [`realtime:subscribe` API endpoint](/api/controllers/realtime/subscribe/index.md). The notifications will be sent to the connection identified by the `Request` object passed to the method.

Reviewer 2: Nitpicking: we try to use the term "action" and not method or endpoint which can be confusing (the first with a method in the code and the second because it's mainly related to HTTP)
```suggestion
Registers a new realtime subscription on behalf of a client. The subscription works exactly like the one created by the [realtime:subscribe](/api/controllers/realtime/subscribe/index.md) API action. The notifications will be sent to the connection identified by the `Request` object passed to the method.
```

---

### Arguments

```js

> subscribe(request)

Reviewer 2: This signature is outdated

---

### Example

```js
async subscribeToSomething(request) {
  const customRequest = new Request(
    {
      action: request.input.action,
      body: {
        equals: {
          myAttribute: 'mycollection'
        }
      },
      collection: 'yellow-taxi',
      controller: request.input.controller,
      index: 'myindex',
    },
    {
      connectionId: request.context.connection.id,
    });

  const options = {}
  const roomId = await this.context.accessors.realtime.subscribe(customRequest);
  return { roomId }
}
```

---

### Use cases

The most common use-case of this method is to implement the realtime subscriptions on the server-side. Instead of leaving the client-side with the freedom (and the responsibility) of specifying the filters for a subscription, you can open a custom API endpoint in your plugin and call `accessors.realtime.subscribe` inside it by passing the incoming Request. When the client hits this endpoint, they will start receiving the notifications from the newly created subscription. This way, the client will only be aware of a `subscribeToSomething` business-specific endpoint, but won't see the set of filters the endpoint encapsulates. 

::: info

> Note that the same restrictions can be achieved by adding a pipe on the `realtime:subscribe` API endpoint. The pipe would check the filters object and make the request fail if the filters are not allowed. This approach has two main downsides:

Reviewer 2: ```suggestion
Note that the same restrictions can be achieved by adding a pipe on the `realtime:subscribe` API action. The pipe would check the filters object and make the request fail if the filters are not allowed. This approach has two main downsides:
```

---
code: true
type: page
title: registerSubscription
---

# `realtime`

> Add or remove [real-time subscriptions](/core/2/guides/essentials/real-time) from the backend.

Reviewer: ```suggestion
Adds or removes [real-time subscriptions](/core/2/guides/essentials/real-time) from the backend.
```

## `registerSubscription`

> Registers a new realtime subscription. The subscription works exaclty like the one created by the [`realtime:subscribe` API endpoint](/api/controllers/realtime/subscribe/index.md). The notifications will be sent to the connection identified by the `Request` object passed to the method.

Reviewer: ```suggestion
Registers a new realtime subscription on behalf of a client. The subscription works exactly like the one created by the [`realtime:subscribe` API endpoint](/api/controllers/realtime/subscribe/index.md). The notifications will be sent to the connection identified by the `Request` object passed to the method.
```

| `request` | `Request` | The request object that contains the subscription payload |

---

### Return

Resolves to the room ID associated with the registered subscription.

---

### Example

```js
async customAction(request) {
  const filters = {
    equals: {
      myAttribute: 'my-value'
    }
  }
  const options = {}
  const roomId = await this.context.accessors.registerSubscription(
        request.context.connection.id,
        'myindex',
        'mycollection',
        filters,
        options
      );
  return { roomId }
}
```

---

### Use cases

The most common use-case of this method is to implement the realtime subscriptions on the server-side. Instead of leaving the client-side with the freedom (and the responsibility) of specifying the filters for a subscription, you can open a custom API endpoint in your plugin and call `registerSubscription` inside it by passing the incoming Request. When the client hits this endpoint, they will start receiving the notifications from the newly created subscription. This way, the client will only be aware of a `subscribeToSomething` business-specific endpoint, but won't see the set of filters the endpoint encapsulates. 

::: info

> Note that the same restriction can be achieved by adding a pipe on the `realtime:subscribe` API endpoint. The pipe would check the filters object and make the request fail if the filters are not allowed. This approach has two main downsides:

Reviewer: ```suggestion
Note that the same restrictions can be achieved by adding a pipe on the `realtime:subscribe` API endpoint. The pipe would check the filters object and make the request fail if the filters are not allowed. This approach has two main downsides:
```
